File: app/src/main/java/org/neidhardt/dynamicsoundboard/soundmanagement/views/SoundSettingsDialog.java
package org.neidhardt.dynamicsoundboard.soundmanagement.views;

import android.annotation.SuppressLint;
import android.app.Dialog;
import android.app.FragmentManager;
import android.net.Uri;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.v7.app.AppCompatDialog;
import android.view.View;
import android.widget.CheckBox;
import android.widget.CompoundButton;
import de.greenrobot.event.EventBus;
import org.neidhardt.dynamicsoundboard.R;
import org.neidhardt.dynamicsoundboard.dao.MediaPlayerData;
import org.neidhardt.dynamicsoundboard.dao.SoundSheet;
import org.neidhardt.dynamicsoundboard.mediaplayer.EnhancedMediaPlayer;
import org.neidhardt.dynamicsoundboard.soundmanagement.dialog.RenameSoundFileDialog;
import org.neidhardt.dynamicsoundboard.soundmanagement.events.SoundChangedEvent;
import org.neidhardt.dynamicsoundboard.views.edittext.CustomEditText;
import org.neidhardt.dynamicsoundboard.views.spinner.CustomSpinner;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

/**
 * File created by eric.neidhardt on 23.02.2015.
 */
public class SoundSettingsDialog extends SoundSettingsBaseDialog implements View.OnClickListener, CompoundButton.OnCheckedChangeListener
{
	private static final String TAG = SoundSettingsDialog.class.getName();

	private CustomEditText soundName;
	private CustomEditText soundSheetName;
	private CustomSpinner soundSheetSpinner;
	private CheckBox addNewSoundSheet;

	private int indexOfCurrentFragment = -1;

	public static void showInstance(FragmentManager manager, MediaPlayerData playerData)
	{
		SoundSettingsDialog dialog = new SoundSettingsDialog();
		addArguments(dialog, playerData.getPlayerId(), playerData.getFragmentTag());
		dialog.show(manager, TAG);
	}

	@NonNull
	@Override
	public Dialog onCreateDialog(Bundle savedInstanceState)
	{
		@SuppressLint("InflateParams") View view = this.getActivity().getLayoutInflater().inflate(R.layout.dialog_sound_settings_layout, null);

		this.soundName = (CustomEditText)view.findViewById(R.id.et_name_file);
		this.soundSheetName = (CustomEditText)view.findViewById(R.id.et_name_new_sound_sheet);
		this.soundSheetSpinner = (CustomSpinner)view.findViewById(R.id.s_sound_sheets);
		this.addNewSoundSheet = (CheckBox)view.findViewById(R.id.cb_add_new_sound_sheet);

		this.addNewSoundSheet.setOnCheckedChangeListener(this);

		view.findViewById(R.id.b_cancel).setOnClickListener(this);
		view.findViewById(R.id.b_ok).setOnClickListener(this);

		this.soundName.setText(this.player.getMediaPlayerData().getLabel());

		this.setAvailableSoundSheets();

		this.soundSheetName.setText(this.getSoundSheetsDataUtil().getSuggestedName());
		this.soundSheetName.setVisibility(View.GONE);

		AppCompatDialog dialog = new AppCompatDialog(this.getActivity(), R.style.DialogTheme);
		dialog.setContentView(view);
		dialog.setTitle(R.string.dialog_sound_settings_title);

		return dialog;
	}

	private void setAvailableSoundSheets()
	{
		List<SoundSheet> soundSheets = this.getSoundSheetsDataAccess().getSoundSheets();
		ArrayList<String> labels = new ArrayList<>();
		for (int i = 0; i < soundSheets.size(); i++)
		{
			if (soundSheets.get(i).getFragmentTag().equals(this.fragmentTag))
				this.indexOfCurrentFragment = i;
			labels.add(soundSheets.get(i).getLabel());
		}
		if (this.indexOfCurrentFragment == -1)
			throw new IllegalStateException(TAG + " Current fragment of sound " + this.player.getMediaPlayerData() + " is not found in list of sound sheets " + soundSheets);

		this.soundSheetSpinner.setItems(labels);
		this.soundSheetSpinner.setSelectedItem(this.indexOfCurrentFragment);
	}

	@Override
	public void onCheckedChanged(CompoundButton buttonView, boolean isChecked)
	{
		if (isChecked)
		{
			this.soundSheetSpinner.setVisibility(View.GONE);
			this.soundSheetName.setVisibility(View.VISIBLE);
		}
		else
		{
			this.soundSheetName.setVisibility(View.GONE);
			this.soundSheetSpinner.setVisibility(View.VISIBLE);
		}
	}

	@Override
	public void onClick(@NonNull View v)
	{
		switch (v.getId())
		{
			case R.id.b_cancel:
				this.dismiss();
				break;
			case R.id.b_ok:
				boolean hasLabelChanged = !this.player.getMediaPlayerData().getLabel().equals(this.soundName.getDisplayedText());
				this.deliverResult();
				this.dismiss();
				if (hasLabelChanged)
					new RenameSoundFileDialog(this.getFragmentManager(), this.player.getMediaPlayerData());
				break;
		}
	}

	private void deliverResult()
	{
		String soundLabel = this.soundName.getDisplayedText();
		int indexOfSelectedSoundSheet = this.soundSheetSpinner.getSelectedItemPosition();

		boolean addNewSoundSheet = this.addNewSoundSheet.isChecked();
		boolean hasSoundSheetChanged = addNewSoundSheet || indexOfSelectedSoundSheet != this.indexOfCurrentFragment;

		if (!hasSoundSheetChanged)
		{
			this.player.getMediaPlayerData().setLabel(soundLabel);
			this.player.getMediaPlayerData().updateItemInDatabaseAsync();
			EventBus.getDefault().post(new SoundChangedEvent(this.player));
		}
		else
		{
			this.getSoundsDataStorage().removeSounds(Collections.singletonList(this.player));

			Uri uri = Uri.parse(this.player.getMediaPlayerData().getUri());

			MediaPlayerData mediaPlayerData;

			if (addNewSoundSheet)
			{
				String soundSheetName = this.soundSheetName.getDisplayedText();
				SoundSheet soundSheet = this.getSoundSheetsDataUtil().getNewSoundSheet(soundSheetName);

				this.getSoundSheetsDataStorage().addSoundSheetToManager(soundSheet);

				String fragmentTag = soundSheet.getFragmentTag();
				mediaPlayerData = EnhancedMediaPlayer.getMediaPlayerData(fragmentTag, uri, soundLabel);
			}
			else
				mediaPlayerData = EnhancedMediaPlayer.getMediaPlayerData(this.fragmentTag, uri, soundLabel);

			this.getSoundsDataStorage().createSoundAndAddToManager(mediaPlayerData);
		}
	}

}

File: app/src/main/kotlin/org/neidhardt/dynamicsoundboard/soundmanagement/views/SoundSettingsDialog.kt
package org.neidhardt.dynamicsoundboard.soundmanagement.views

import android.annotation.SuppressLint
import android.app.Dialog
import android.net.Uri
import android.os.Bundle
import android.support.v4.app.FragmentManager
import android.support.v7.app.AlertDialog
import android.view.View
import android.widget.CheckBox
import android.widget.CompoundButton
import org.greenrobot.eventbus.EventBus
import org.neidhardt.dynamicsoundboard.R
import org.neidhardt.dynamicsoundboard.SoundboardApplication
import org.neidhardt.dynamicsoundboard.dao.MediaPlayerData
import org.neidhardt.dynamicsoundboard.mediaplayer.MediaPlayerController
import org.neidhardt.dynamicsoundboard.soundmanagement.dialog.RenameSoundFileDialog
import org.neidhardt.dynamicsoundboard.soundmanagement.events.SoundChangedEvent
import org.neidhardt.dynamicsoundboard.views.edittext.CustomEditText
import org.neidhardt.dynamicsoundboard.views.spinner.CustomSpinner
import java.util.*

/**
 * File created by eric.neidhardt on 23.02.2015.
 */
class SoundSettingsDialog : SoundSettingsBaseDialog(), CompoundButton.OnCheckedChangeListener {

	private val soundSheetsDataUtil = SoundboardApplication.soundSheetsDataUtil
	private val soundsDataStorage = SoundboardApplication.soundsDataStorage
	private val soundSheetsDataAccess = SoundboardApplication.soundSheetsDataAccess
	private val soundSheetsDataStorage = SoundboardApplication.soundSheetsDataStorage

	override var fragmentTag: String? = null
	override var player: MediaPlayerController? = null

	private var soundName: CustomEditText? = null
	private var soundSheetName: CustomEditText? = null
	private var soundSheetSpinner: CustomSpinner? = null
	private var addNewSoundSheet: CheckBox? = null

	private var indexOfCurrentFragment = -1

	override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
		@SuppressLint("InflateParams") val view = this.activity.layoutInflater.inflate(R.layout.dialog_sound_settings_layout, null)

		this.soundName = view.findViewById(R.id.et_name_file) as CustomEditText
		this.soundSheetName = view.findViewById(R.id.et_name_new_sound_sheet) as CustomEditText
		this.soundSheetSpinner = view.findViewById(R.id.s_sound_sheets) as CustomSpinner
		this.addNewSoundSheet = view.findViewById(R.id.cb_add_new_sound_sheet) as CheckBox

		this.addNewSoundSheet!!.setOnCheckedChangeListener(this)

		this.soundName!!.text = this.player!!.mediaPlayerData.label

		this.setAvailableSoundSheets()

		this.soundSheetName!!.text = this.soundSheetsDataUtil.getSuggestedName()
		this.soundSheetName!!.visibility = View.GONE

		return AlertDialog.Builder(context).apply {
			this.setTitle(R.string.dialog_sound_settings_title)
			this.setView(view)
			this.setPositiveButton(R.string.dialog_save, { dialogInterface, i ->
				val hasLabelChanged = player!!.mediaPlayerData.label != soundName!!.displayedText
				deliverResult()
				dismiss()
				if (hasLabelChanged)
					RenameSoundFileDialog(fragmentManager, player!!.mediaPlayerData)
			})
			this.setNegativeButton(R.string.dialog_cancel, { dialogInterface, i -> dismiss()})
		}.create()
	}

	private fun setAvailableSoundSheets() {
		val soundSheets = this.soundSheetsDataAccess.getSoundSheets()
		val labels = ArrayList<String>()
		for (i in soundSheets.indices) {
			if (soundSheets[i].fragmentTag == this.fragmentTag)
				this.indexOfCurrentFragment = i
			labels.add(soundSheets[i].label)
		}
		if (this.indexOfCurrentFragment == -1)
			throw IllegalStateException(TAG + " Current fragment of sound " + this.player!!.mediaPlayerData + " is not found in list of sound sheets " + soundSheets)

		this.soundSheetSpinner!!.setItems(labels)
		this.soundSheetSpinner!!.setSelectedItem(this.indexOfCurrentFragment)
	}

	override fun onCheckedChanged(buttonView: CompoundButton, isChecked: Boolean) {
		if (isChecked) {
			this.soundSheetSpinner!!.visibility = View.GONE
			this.soundSheetName!!.visibility = View.VISIBLE
		} else {
			this.soundSheetName!!.visibility = View.GONE
			this.soundSheetSpinner!!.visibility = View.VISIBLE
		}
	}

	private fun deliverResult() {
		val soundLabel = this.soundName!!.displayedText
		val indexOfSelectedSoundSheet = this.soundSheetSpinner!!.selectedItemPosition

		val addNewSoundSheet = this.addNewSoundSheet!!.isChecked
		val hasSoundSheetChanged = addNewSoundSheet || indexOfSelectedSoundSheet != this.indexOfCurrentFragment

		if (!hasSoundSheetChanged) {
			this.player!!.mediaPlayerData.label = soundLabel
			this.player!!.mediaPlayerData.updateItemInDatabaseAsync()
			EventBus.getDefault().post(SoundChangedEvent(this.player!!))
		} else {
			this.soundsDataStorage.removeSounds(listOf(this.player!!))

			val uri = Uri.parse(this.player?.mediaPlayerData?.uri)

			val mediaPlayerData: MediaPlayerData

			if (addNewSoundSheet) {
				val soundSheetName = this.soundSheetName!!.displayedText
				val soundSheet = this.soundSheetsDataUtil.getNewSoundSheet(soundSheetName)

				this.soundSheetsDataStorage.addSoundSheetToManager(soundSheet)

				val fragmentTag = soundSheet.fragmentTag
				mediaPlayerData = MediaPlayerData.getNewMediaPlayerData(fragmentTag, uri, soundLabel)
			} else
				mediaPlayerData = MediaPlayerData.getNewMediaPlayerData(this.fragmentTag, uri, soundLabel)

			this.soundsDataStorage.createSoundAndAddToManager(mediaPlayerData)
		}
	}

	companion object {
		private val TAG = SoundSettingsDialog::class.java.name

		fun showInstance(manager: FragmentManager, playerData: MediaPlayerData) {
			val dialog = SoundSettingsDialog()
			SoundSettingsBaseDialog.Companion.addArguments(dialog, playerData.playerId, playerData.fragmentTag)
			dialog.show(manager, TAG)
		}
	}

}
